enhancements/sparse-manifest-list.md

```diff
@@ -0,0 +1,111 @@
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+> 1. What are the potential impacts on Quay's existing storage backend when implementing sparse manifest lists?
+> 2. How will this enhancement affect existing Quay users, and what migration steps are necessary?
+> 3. What changes will be required in client software to be compatible with handling sparse manifest lists? This includes Podman and OC Mirror
+
+## Summary
+
+The introduction of sparse manifest list support in Quay container registry aims to increase the efficiency of how container images are stored and managed. This enhancement focuses on optimizing the use of storage space and minimizing the data transfer when pulling images from the registry. The feature will be particularly beneficial for users with large image repositories or those operating in bandwidth-constrained environments.
+
+## Motivation
+
+With the growing size of container images and the increasing need for efficiency in container registry operations, there's a clear need for more advanced storage solutions. Sparse manifest lists offer a compelling approach to address these challenges. This enhancement seeks to leverage these benefits to improve Quay's functionality and user experience. This will also help reduce costs on Openshift in supporting OCP installations in various architectures.
+
+### Goals
+
+- Implement sparse manifest list support in Quay container registry.
+- Optimize storage utilization and data transfer efficiency.
+- Ensure backward compatibility and minimal disruption for current users.
+- Ensure that existing clients are able to handle pulling these images, and that other clients that do not support this feature necessarily are not broken
+
+### Non-Goals
+
+- Overhaul of Quay's existing storage architecture.
+- Disruption of existing workflows for current Quay users.
+
+## Proposal
+
+The proposal includes developing and integrating sparse manifest list support into the Quay container registry. This will involve changes in how Quay stores and handles container images, focusing on layer deduplication and efficient data management.
+
+### User Stories [optional]
+
+#### Story 1
+
+A user with a large repository of container images can significantly reduce their storage footprint by using sparse manifest lists, as common layers across different images are stored only once.
+
+#### Story 2
+
+In a bandwidth-constrained environment, a user can pull images from Quay more efficiently, as the sparse manifest list allows downloading only the necessary layers, reducing the data transfer volume.
```

Comment:
Clients can pull a portion of an image already, e.g. see [this comment][1] in openshift/oc#1334 adding sparse manifest support to `oc image mirror ...`.  And I agree with [Miloslav's comment][2] about "common layers" seeming orthogonal.  My understanding is that the user-story for sparse manifest lists is more like:

> Many image authors publish manifest-list images with many architecture-specific children, to support their workload on all of those architectures.  Some image consumers only run a subset of those architectures locally.  Sparse manifests will allow users mirroring a manifest-list image into their local Quay to only push the architectures they need, while retaining the top-level manifest list.  This saves the network bandwidth and local-Quay storage costs of mirroring architectures that are not needed locally.  And it preserves the digest and signatures on the original manifest-list.

So for:

```console
$ curl -s https://quay.io/v2/openshift-release-dev/ocp-release/manifests/sha256:39aa3985a4ab715f3ea8d983b72745947249322e4fb4dbcf59b4cc749f4e9ae7 | jq -r '.manifests[] | .digest + " " + (.platform | tostring)'
sha256:49821163426f2f2cb5a2b7cb446c35440d6a5c3905397b48b795dd4bc3b5eaf6 {"architecture":"amd64","os":"linux"}
sha256:f00ca1a7bef6176803cd54ad8ae878dd48fa86215dd002b834840f01039de045 {"architecture":"ppc64le","os":"linux"}
sha256:99696da77b6982057442bdba3854ddd574e5aeba6bd1710e138b8b398b22f883 {"architecture":"s390x","os":"linux"}
sha256:a6352c78572180f0e88cbf62f80f7b45074a157d4e3d8ad172e7d77042f06724 {"architecture":"arm64","os":"linux"}
```

The `sha256:39aa398...` manifest-list would be pushed into the local Quay, along with `sha256:4982116....` amd64 and `sha256:a6352c78572...` arm64.  But `sha256:f00ca1a7bef61...` ppc64le and `sha256:99696da77b69...` s390x would not be pushed in.  As far as Quay-side changes go, that's almost entirely on the what-can-we-push-into-Quay? side and not on the what-can-we-pull-from-Quay? side.

As [Miloslav points out][3], clients who are pulling from Quay and expecting a full image but receiving a sparse manifest will fail to pull the layers ("hey, this manifest list references `sha256:f00ca1a7bef61...`, but that is 404ing!").  They'd have to sort that out with some kind knob.  And as Miloslav points out, Quay-to-Quay mirroring would also have to handle the source-manifest-list-is-sparse case.

[1]: https://github.com/openshift/oc/pull/1334#discussion_r1149468369
[2]: https://github.com/quay/enhancements/pull/29#discussion_r1499639222
[3]: https://github.com/quay/enhancements/pull/29#issuecomment-1959976190